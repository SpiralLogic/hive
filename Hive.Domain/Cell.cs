using System.Collections.Generic;

namespace Hive.Domain
{
    public class Cell
    {
        public Cell(GameCoordinate position, string color)
        {
            Position = position;
            Color = color;
        }

        public GameCoordinate Position { get; }
        public string Color { get; }
        public ICollection<Tile> Tiles { get; } = new List<Tile>();
    }
}