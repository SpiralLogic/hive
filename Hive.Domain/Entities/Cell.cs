﻿using System.Collections.Generic;

namespace Hive.Domain.Entities
{
    public class Cell
    {
        public Cell(GameCoordinate coordinates, ICollection<Tile> tiles = null)
        {
            Coordinates = coordinates;
            Tiles = tiles ??  new List<Tile>();
        }

        public GameCoordinate Coordinates { get; }
        public ICollection<Tile> Tiles { get; }
    }
}