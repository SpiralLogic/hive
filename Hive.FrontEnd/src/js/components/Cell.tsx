import * as React from 'react';
import { useDrop } from 'react-dnd';
import { Hexagon, HexCoordinates } from '../domain';
import { Tile, TILE_TYPE } from './Tile';

type CellProps = Hexagon

const areEqual = (a: HexCoordinates, b: HexCoordinates) => a && b && a.q === b.q && a.r === b.r;
const isValidMove = (move: HexCoordinates, validMoves: HexCoordinates[]) => validMoves.some(dest => areEqual(move, dest));
const coordinateToString = ({ q, r }: HexCoordinates) => `${q}-${r}`;

export const Cell: React.FunctionComponent<CellProps> = ({
    tiles,
    coordinates,
}) => {
    const [{ isOver, canDrop }, drop] = useDrop({
        accept: TILE_TYPE,
        drop: () => coordinates,
        canDrop: (item, monitor) => isValidMove(coordinates, monitor.getItem().availableMoves),
        collect: monitor => ({
            isOver: monitor.isOver(),
            canDrop: monitor.canDrop(),
        }),
    });

    const backGroundColor = () => {
        if (canDrop && isOver) return ' valid-cell active';
        if (isOver) return ' invalid-cell';
        if (canDrop) return ' valid-cell';
        return '';
    };
    const id = coordinateToString(coordinates);
    return (
        <div className={'hex cell' + backGroundColor()} ref={drop}>
            {tiles.length > 0 && <Tile key={'tile' + id} {...tiles[0]} />}
        </div>
    );
};

Cell.displayName = 'Cell';