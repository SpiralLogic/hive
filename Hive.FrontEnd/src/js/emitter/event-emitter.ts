export type EventEmitter<Event> = {
    emit: (event: Event) => void;
    add: (...listeners: EventListener<Event>[]) => void;
    remove: (...listeners: EventListener<Event>[]) => void;
};

export type EventListener<E> = (event: E) => void;

export type Event = { type: string };
