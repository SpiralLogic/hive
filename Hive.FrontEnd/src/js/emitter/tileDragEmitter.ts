import {HexCoordinates, TileId} from '../domain';
import {EventEmitter, EventListener} from './interface';

export type TileDragListener = EventListener<TileDragEvent>;

export type TileDragEvent = {
    type: 'start' | 'end',
    source: TileId,
    data: HexCoordinates[]
}

export default class TileDragEmitter implements EventEmitter<TileDragEvent> {
    private listeners = new Set<TileDragListener>();

    add(...listeners: TileDragListener[]) {
        listeners.forEach(l => this.listeners.add(l));
    }

    emit(event: TileDragEvent) {
        this.listeners.forEach(l => l(event));
    }

    remove(listener: TileDragListener) {
        this.listeners.delete(listener);
    }
}