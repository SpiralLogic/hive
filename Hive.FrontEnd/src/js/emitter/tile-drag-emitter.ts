import { HexCoordinates, TileId } from '../domain';
import { EventEmitter, EventListener } from './event-emitter';

export type TileDragListener = EventListener<TileDragEvent>;

export type TileDragEvent = {
    type: 'start' | 'end' | 'enter' | 'leave';
    tileId: TileId;
    tileMoves: HexCoordinates[];
};

export default class TileDragEmitter implements EventEmitter<TileDragEvent> {
    private listeners = new Set<TileDragListener>();

    add(...listeners: TileDragListener[]) {
        listeners.forEach((l) => this.listeners.add(l));
    }

    emit(event: TileDragEvent) {
        this.listeners.forEach((l) => l(event));
    }

    remove(...listeners: TileDragListener[]) {
        listeners.forEach((l) => this.listeners.delete(l));
    }
}
