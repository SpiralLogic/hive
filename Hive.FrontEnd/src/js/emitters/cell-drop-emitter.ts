import { HexCoordinates, TileId } from '../domain';
import { EventEmitter } from './event-emitter';

export class CellDropEmitter extends EventEmitter<CellDropEvent> {
}

export type CellDropListener<CellDropEvent> = (event: CellDropEvent) => void;

const cellDropEmitter = new CellDropEmitter();

export type CellDropEvent = {
    type: 'drop';
    coords: HexCoordinates;
    tileId: TileId;
};

export const useCellDropEmitter = () => cellDropEmitter;
