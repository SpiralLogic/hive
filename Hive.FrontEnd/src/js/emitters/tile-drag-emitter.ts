import { HexCoordinates, TileId } from '../domain';
import { EventEmitter } from './event-emitter';

export type TileDragEvent = {
    type: 'start' | 'end';
    tileId: TileId;
    moves: HexCoordinates[];
};

export class TileDragEmitter extends EventEmitter<TileDragEvent> {
}

const tileDragEmitter = new TileDragEmitter();

export const useTileDragEmitter = () => tileDragEmitter;
