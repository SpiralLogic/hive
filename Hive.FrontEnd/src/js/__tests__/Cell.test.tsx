import { act, fireEvent, render } from '@testing-library/preact';
import Cell from '../components/Cell';
import { h } from 'preact';
import { useCellDropEmitter } from '../emitter/cell-drop-emitter';
import { useTileDragEmitter } from '../emitter/tile-drag-emitter';
import { deepEqual } from 'fast-equals';
import { simulateEvent } from './helpers/helpers';

jest.mock('fast-equals', () => ({ deepEqual: jest.fn(() => true) }));

const moveTileSpy = jest.fn();

const cellWithNoTile = () => {
    const cell = { coordinates: { q: 0, r: 0 }, tiles: [], moveTile: moveTileSpy };
    return render(<Cell {...cell}/>).container.firstElementChild as HTMLElement;
};

const cellWithTile = () => {
    const tile = { id: 2, playerId: 2, content: 'fly', availableMoves: [] };
    const cell = { coordinates: { q: 1, r: 1 }, tiles: [tile], moveTile: moveTileSpy };

    return render(<Cell {...cell}/>).container.firstElementChild as HTMLElement;
};

const canDropCellWithTile = () => {
    const tile = { id: 2, playerId: 2, content: 'ant', availableMoves: [] };
    const cell = { coordinates: { q: 2, r: 2 }, tiles: [tile], moveTile: moveTileSpy };

    return render(<Cell {...cell}/>).container.firstElementChild as HTMLElement;
};

const noDropEmptyCell = () => {
    const cell = { coordinates: { q: 0, r: 0 }, tiles: [], moveTile: moveTileSpy };
    return render(<Cell {...cell}/>).container.firstElementChild as HTMLElement;
};

const canDropEmptyCell = cellWithNoTile;
const noDropCellWithTile = cellWithTile;

describe('Cell Render', () => {
    test('starts with default classes', () => {
        expect(cellWithTile().classList.value).toEqual('hex cell');
        expect(cellWithNoTile().classList.value).toEqual('hex cell');
    });

    test('top tile is rendered if it exists', () => {
        const tiles = cellWithTile().getElementsByClassName('tile');
        expect(tiles).toHaveLength(1);
    });

    test('Cell is memoized with deep equal', () => {
        const props = { coordinates: { q: 0, r: 0 }, tiles: [], moveTile: moveTileSpy };
        const cell = <Cell {...props}/>;
        render(cell).rerender(cell);
        expect(deepEqual).toHaveBeenCalledTimes(1);
    });
});

describe('Cell drag and drop', () => {
    const emitter = useTileDragEmitter();

    function emitTileEvent (type: 'start' | 'end') {
        act(() => emitter.emit({ type, tileId: 2, tileMoves: [{ q: 0, r: 0 }, { q: 2, r: 2 }] }));
    }

    test('dragover allows drop', () => {
        const preventDefault = simulateEvent(cellWithTile(), 'dragover');

        expect(preventDefault).toHaveBeenCalled();
    });

    test('cell is valid on drag start', () => {
        const cellWithTile = canDropCellWithTile();
        const emptyCell = canDropEmptyCell();
        emitTileEvent('start');

        expect(cellWithTile).toHaveClass('can-drop');
        expect(emptyCell).toHaveClass('can-drop');
    });

    test('valid cell is active on tile drag enter', () => {
        const cellWithTile = canDropCellWithTile();
        const emptyCell = canDropEmptyCell();
        emitTileEvent('start');
        fireEvent.dragEnter(cellWithTile);
        fireEvent.dragEnter(emptyCell);

        expect(cellWithTile).toHaveClass('active');
        expect(emptyCell).toHaveClass('active');
    });

    test('drop sends move tile request when valid and active', () => {
        jest.spyOn(useCellDropEmitter(), 'emit');
        const cellWithTile = canDropCellWithTile();
        const emptyCell = canDropEmptyCell();
        emitTileEvent('start');
        fireEvent.dragEnter(cellWithTile);
        fireEvent.dragEnter(emptyCell);
        emitTileEvent('end');

        expect(useCellDropEmitter().emit).toHaveBeenCalledWith({ type: 'drop', tileId: 2, coordinates: { q: 0, r: 0 } });
        expect(useCellDropEmitter().emit).toHaveBeenCalledWith({type: 'drop', tileId: 2, coordinates: { q: 2, r: 2 } });
    });

    test('drop doesnt send move tile request when cell is no drop', () => {
        jest.spyOn(useCellDropEmitter(), 'emit');
        noDropCellWithTile();
        noDropEmptyCell();
        emitTileEvent('start');
        emitTileEvent('end');

        expect(useCellDropEmitter().emit).not.toHaveBeenCalled();
    });

    test('invalid cell doesnt send move request', () => {
        noDropCellWithTile();
        noDropEmptyCell();
        document.querySelectorAll('.cell').forEach(c => fireEvent.dragEnter(c));

        expect(moveTileSpy).not.toHaveBeenCalled();
    });

    test('active and no-drop are removed on drag leave', () => {
        cellWithTile();
        cellWithNoTile();
        noDropCellWithTile();
        noDropEmptyCell();
        emitTileEvent('start');
        document.querySelectorAll('.cell').forEach(c => fireEvent.dragEnter(c));
        document.querySelectorAll('.cell').forEach(c => fireEvent.dragLeave(c));

        expect(document.getElementsByClassName('active')).toHaveLength(0);
        expect(document.getElementsByClassName('no-drop')).toHaveLength(0);
    });
});

describe('Cell Snapshot', () => {
    test('cell with tile matches current snapshot', () => {
        expect(cellWithTile()).toMatchSnapshot();
    });

    test('cell with no tile matches current snapshot', () => {
        expect(cellWithNoTile()).toMatchSnapshot();
    });
});
