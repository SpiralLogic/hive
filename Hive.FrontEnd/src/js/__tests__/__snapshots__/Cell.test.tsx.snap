// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell Snapshot matches current snapshot 1`] = `
<body>
  <div>
    <div
      class="hex cell"
    />
    <div
      class="hex cell"
    >
      <div
        class="hex tile"
        draggable="true"
        style="--color: red;"
        title="fly"
      >
        <span>
          fly
        </span>
      </div>
    </div>
  </div>
  <div>
    <div
      class="hex cell"
    />
    <div
      class="hex cell"
    >
      <div
        class="hex tile"
        draggable="true"
        style="--color: red;"
        title="fly"
      >
        <span>
          fly
        </span>
      </div>
    </div>
  </div>
</body>
`;
